go-runtime/compile/build.go

```diff
@@ -978,23 +975,128 @@ func goVerbFromQualifiedName(qualifiedName string) (goVerb, error) {
 	}, nil
 }
 
-// package and directory names are the same (dir=bar, pkg=bar): "github.com/foo/bar.A" => "github.com/foo/bar"
-// package and directory names differ (dir=bar, pkg=baz): "github.com/foo/bar.baz.A" => "baz github.com/foo/bar"
-func goImportFromQualifiedName(qualifiedName string) (string, error) {
+// returns the import path and directory name for an external type
+// package and directory names are the same (dir=bar, pkg=bar): "github.com/foo/bar.A" => "github.com/foo/bar", none
+// package and directory names differ (dir=bar, pkg=baz): "github.com/foo/bar.baz.A" => "github.com/foo/bar", "baz"
+func goImportFromQualifiedName(qualifiedName string) (importPath string, directoryName optional.Option[string], err error) {
 	lastDotIndex := strings.LastIndex(qualifiedName, ".")
 	if lastDotIndex == -1 {
-		return "", fmt.Errorf("invalid qualified type format %q", qualifiedName)
+		return "", optional.None[string](), fmt.Errorf("invalid qualified type format %q", qualifiedName)
 	}
 
 	pkgPath := qualifiedName[:lastDotIndex]
 	pkgName := path.Base(pkgPath)
 
-	importAlias := ""
 	if lastDotIndex = strings.LastIndex(pkgName, "."); lastDotIndex != -1 {
 		pkgName = pkgName[lastDotIndex+1:]
 		pkgPath = pkgPath[:strings.LastIndex(pkgPath, ".")]
-		// package and path differ, so we need to alias the import
-		importAlias = pkgName + " "
+		return pkgPath, optional.Some(pkgName), nil
+	}
+	return pkgPath, optional.None[string](), nil
+}
+
+// imports returns a map of import paths to aliases for a module.
+// - hardcoded for time ("stdtime")
+// - prefixed with "ftl" for other modules (eg "ftlfoo")
+// - addImports() is used to generate shortest unique aliases for external packages
+func imports(m *schema.Module, aliasesMustBeExported bool) map[string]string {
+	// find all imports
+	imports := map[string]string{}
+	// map from import path to the first dir we see
+	extraImports := map[string]optional.Option[string]{}
+	_ = schema.VisitExcludingMetadataChildren(m, func(n schema.Node, next func() error) error { //nolint:errcheck
+		switch n := n.(type) {
+		case *schema.Ref:
+			if n.Module == "" || n.Module == m.Name {
+				break
+			}
+			imports[path.Join("ftl", n.Module)] = "ftl" + n.Module
+			for _, tp := range n.TypeParameters {
+				if tpRef, ok := tp.(*schema.Ref); ok && tpRef.Module != "" && tpRef.Module != m.Name {
+					imports[path.Join("ftl", tpRef.Module)] = "ftl" + tpRef.Module
+				}
+			}
+
+		case *schema.Time:
+			imports["time"] = "stdtime"
+
+		case *schema.Optional, *schema.Unit:
+			imports["github.com/TBD54566975/ftl/go-runtime/ftl"] = ""
+
+		case *schema.Topic:
+			if n.IsExported() {
+				imports["github.com/TBD54566975/ftl/go-runtime/ftl"] = ""
+			}
+
+		case *schema.TypeAlias:
+			if !aliasesMustBeExported || n.IsExported() {
```

Comment:
I am not sure about the context here, so I'm probably wrong here, but thought I'd flag it in case.

Reading this if statement makes me think this block is for exported aliases, but its checking for when aliases must not be exported? Should it be `aliasesMustBeExported || n.IsExported()`?

Reply:
I think what I want is:
```
if aliasesMustBeExported && !n.IsExported() {
  // early exit
}
```
When i wrote the original line, it also struck me as odd / confusing, so I'll replace it with the above as it's a lot clearer